util/rlp/src/stream.rs

```diff
@@ -0,0 +1,245 @@
+// Copyright 2015-2017 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::borrow::Borrow;
+use elastic_array::{ElasticArray16, ElasticArray1024};
+use bytes::{ToBytes, VecLike};
+use traits::RlpEncodable;
+use Stream;
+
+#[derive(Debug, Copy, Clone)]
+struct ListInfo {
+	position: usize,
+	current: usize,
+	max: usize,
+}
+
+impl ListInfo {
+	fn new(position: usize, max: usize) -> ListInfo {
+		ListInfo {
+			position: position,
+			current: 0,
+			max: max,
+		}
+	}
+}
+
+/// Appendable rlp encoder.
+pub struct RlpStream {
+	unfinished_lists: ElasticArray16<ListInfo>,
+	buffer: ElasticArray1024<u8>,
+	finished_list: bool,
+}
+
+impl Default  for RlpStream {
+	fn default() -> Self {
+		RlpStream::new()
+	}
+}
+
+impl RlpStream {
+	pub fn new() -> Self {
+		RlpStream {
+			unfinished_lists: ElasticArray16::new(),
+			buffer: ElasticArray1024::new(),
+			finished_list: false,
+		}
+	}
+
+	pub fn new_list(len: usize) -> Self {
+		let mut stream = RlpStream::new();
+		stream.begin_list(len);
+		stream
+	}
+
+	pub fn append<'a, E>(&'a mut self, value: &E) -> &'a mut Self where E: RlpEncodable {
```

Comment:
One nice change we could make would be to `impl` on `[u8]` instead of `&'a [u8]` (likewise for str), and then do `E: RlpEncodable + ?Sized` so we don't have to double-reference slices (which currently leads to lots of `&&*my_vec` in the codebase)